### PR TITLE
BUG: value_counts returning incorrect dtype for string dtype

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -2855,7 +2859,9 @@ def _value_counts(
             result_series.name = name
             result_series.index = index.set_names(range(len(columns)))
             result_frame = result_series.reset_index()
-            result_frame.columns = columns + [name]
+            result_frame.columns = Index(
+                columns + [name], dtype=self.grouper.groupings[0].obj.columns.dtype
```
Comment (phofl) [inline pandas/core/groupby/groupby.py:2863]: @rhshadrach is there a better way of getting the dtype of the original column index?

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:2863]: self.obj.columns.dtype I think.

Comment (phofl) [reply pandas/core/groupby/groupby.py:2863]: That's a Series in SeriesGroupBy unfortunately

**Review by rhshadrach — CHANGES_REQUESTED** [review]
looks good - I think the size change could use a test.

**Comment by phofl** [comment]
added a test

File: pandas/tests/groupby/methods/test_size.py [hunk 2]
```diff
@@ -104,3 +106,24 @@ def test_size_series_masked_type_returns_Int64(dtype):
...
+@pytest.mark.parametrize(
+    "dtype",
+    [
+        object,
+        pytest.param("string[pyarrow_numpy]", marks=td.skip_if_no("pyarrow")),
+        pytest.param("string[pyarrow]", marks=td.skip_if_no("pyarrow")),
+    ],
+)
+def test_size_strings(dtype):
+    df = DataFrame({"a": ["a", "a", "b"], "b": "a"}, dtype=dtype)
```
Comment (rhshadrach) [inline pandas/tests/groupby/methods/test_size.py:120]: nit: GH#

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks - just missing the GH#

File: pandas/tests/groupby/methods/test_size.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/groupby/methods/test_size.py:121]: added

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @phofl

**Comment by phofl** [comment]
Sorry for the confusion, this should go into 2.1.2 (it#s only really relevant if you use the new option)

**Comment by phofl** [comment]
@meeseeksdev backport 2.1.x

**Pull request merged**